libs/core/tests/unit/mutex.cpp

```diff
@@ -116,4 +119,49 @@ TEST(DebugMutex, DISABLED_MultiThreadDeadlock2)
   threads.clear();
 }
 
+TEST(DebugMutex, Recursive)
+{
+  fetch::MutexRegister::ThrowOnDeadlock();
+  RMutex m;
+
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  m.unlock();
+  m.unlock();
+  m.unlock();
+  m.unlock();
+
+  std::string rv;
+
+  auto f = [&m, &rv](char c) {
+    m.lock();
+    m.lock();
+    m.lock();
+    m.lock();
+
+    rv += c;
+    rv += c;
+    rv += c;
+    rv += c;
+
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+
+    m.unlock();
+    m.unlock();
+    m.unlock();
+    m.unlock();
+  };
+
+  std::vector<std::thread> threads;
+  threads.emplace_back(f, 'a');
+  threads.emplace_back(f, 'b');
+
+  threads[0].join();
+  threads[1].join();
+
+  ASSERT_TRUE(rv == "aaaabbbb" || rv == "bbbbaaaa");
```

Comment:
This test does not seem to test that recursive mutex detects a deadlock/infinite loop. Does recursive mutex throw if there is a deadlock, and on what condition will it throw?

Reply:
Yes, that should be a separate test, as I thought this morning. This one only tests for the lock to be acquired and released recursively, when no deadlock occurs.